### PR TITLE
Building model plugins as sidecar containers

### PR DESCRIPTION
**Pull request by tomikazi**
Still WIP!

The config-models main program has been renamed to `model-compiler` and the associated Dockerfile has been repurposed to generate a model compiler image. This image can be invoked with a volume mount that is expected to contain the model `metadata.yaml` file and directory `yang/` that contains the input YANG files. The image will generate Go API bindings using the ygot generator utility, YANG model tree using the `pyang` utility and it will generate a set of artifacts (Go files, Makefile, go.mod and Dockerfile) for building the plugin executable. The plugin docker image can then be build by simply invoking `make` from the directory that was provided as the volume mount. Example:

`docker run -v $(pwd)/models/devicesim-1.0.0:/config-model onosproject/model-compiler:latest`

When the resulting model plugin image is run, it will start the plugin which will setup the PluginService gRPC endpoint on the local host and the specified port (port number still to be made configurable). The onos-config container can then contact this plugin service to extract model metadata and to register the plugin appropriately. (still WIP)

### CODE REVIEW

**Review by SeanCondon — COMMENTED** [review]
@tomikazi I think you can leave out `aether-3.0.0`, `testdevice-1.0.0`, `testdevice-2.0.0` and `devicesim-1.0.0`  from the deprecated plugins, since `aether-3.0.0` was dure for removal anyway, and the others are already available as go releases in history, and will be used from the new location for any future versions

**Comment by tomikazi** [comment]
> @tomikazi I think you can leave out `aether-3.0.0`, `testdevice-1.0.0`, `testdevice-2.0.0` and `devicesim-1.0.0` from the deprecated plugins, since `aether-3.0.0` was dure for removal anyway, and the others are already available as go releases in history, and will be used from the new location for any future versions

Will do! Thanks.

**Comment by SeanCondon** [comment]
@tomikazi One other thing I though of that you can do in the compiler is to lint the YANG. Here's what we do in Aether:
```
pyang --lint --lint-ensure-hyphenated-names -W error *.yang
```

**Comment by tomikazi** [comment]
> `docker run -v $(pwd)/models/devicesim-1.0.0:/config-model onosproject/model-compiler:latest`

Perfect. I'll add that as well.

File: templates/paths.go.tpl [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+	Enum        map[int]string
+	IsAKey      bool
+	AttrName    string
+}
+
+// ReadOnlySubPathMap abstracts the read only subpath
+type ReadOnlySubPathMap map[string]ReadOnlyAttrib
+
+// ReadOnlyPathMap abstracts the read only path
+type ReadOnlyPathMap map[string]ReadOnlySubPathMap
```
Comment (adibrastegarnia) [inline templates/paths.go.tpl:53]: Can we have a function to form ReadOnlyPathMap and ReadWitePathMap?

**Review by adibrastegarnia — APPROVED** [review]

**Pull request merged**